src/targets/gpu/hip.cpp

```diff
@@ -74,9 +74,15 @@ hip_ptr allocate_gpu(std::size_t sz, bool host = false)
 
 hip_host_ptr register_on_gpu(void* ptr, std::size_t sz)
 {
-    auto status = hipHostRegister(ptr, sz, hipHostRegisterMapped);
+    void* dev_ptr;
+    // check if ptr was allocated through hipHostMalloc
+    auto status = hipHostGetDevicePointer(&dev_ptr, ptr, 0);
```

Comment:
This function returns 3 possible statuses. Should we be concerned with which error status is reported?

Reply:
The other two error codes are 1. hipErrorInvalidValue which gets thrown whenever input `ptr` hipHostGetDevicePointer is not valid.  This is the error code that I get if try to fetch device pointer from host pointer that is not registered i.e. it is invalid. 

Second error code `hipErrorOutofMemory`, I am not sure in what situation it gets thrown for `hipHostGetDevicePointer`. Upon trying to find all references in ROCm software organization for that error code, it seems, it is used when `malloc`  fails.  

I changed the code and added explicit error status checks to be more precise.